# Main Dataframes Cleaning and Wrangling

# Contents

## 01. Set up environment

## 02. gsearch_st data wrangling (cleaning and consistency checks)

## 03. song_info data wrangling (cleaning and consistency checks)

## 04. Exporting cleaned data sets

# 01. Set up environment

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Import visualization libraries

import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
# Create path

path = r"C:\Users\leobs\OneDrive\Desktop\LF_TS_Analysis"

In [1]:
# Import original google search trends by state dataframe

df_gsearch_st = pd.read_csv(os.path.join(path,'01 Sourced Data', 'Main Dataframes', 'google_trends_state_album_release_period_complete.csv'), index_col = False)

NameError: name 'pd' is not defined

In [5]:
# Import original song information dataframe

df_song_info = pd.read_csv(os.path.join(path,'01 Sourced Data', 'Main Dataframes', 'jp_taylor_swift_spotify_sound_charact_2023_10_27.csv'), index_col = False)

# 02. Google search dataframe cleaning and wrangling

## Check columns

In [6]:
# Check dataframe import

df_gsearch_st.head(10)

period_state                 period  \
0               2006/09/28-2006/10/10, Alabama  2006/09/28-2006/10/10   
1                2006/09/28-2006/10/10, Alaska  2006/09/28-2006/10/10   
2               2006/09/28-2006/10/10, Arizona  2006/09/28-2006/10/10   
3              2006/09/28-2006/10/10, Arkansas  2006/09/28-2006/10/10   
4            2006/09/28-2006/10/10, California  2006/09/28-2006/10/10   
5              2006/09/28-2006/10/10, Colorado  2006/09/28-2006/10/10   
6           2006/09/28-2006/10/10, Connecticut  2006/09/28-2006/10/10   
7              2006/09/28-2006/10/10, Delaware  2006/09/28-2006/10/10   
8  2006/09/28-2006/10/10, District of Columbia  2006/09/28-2006/10/10   
9               2006/09/28-2006/10/10, Florida  2006/09/28-2006/10/10   

  period_start_date                 state  g_search_score album_event  \
0        2006-09-28               Alabama               0       Debut   
1        2006-09-28                Alaska               0       Debut   
2        2006-09-28               Arizona               8       Debut   
3        2006-09-28              Arkansas              46       Debut   
4        2006-09-28            California               3       Debut   
5        2006-09-28              Colorado               7       Debut   
6        2006-09-28           Connecticut               0       Debut   
7        2006-09-28              Delaware               0       Debut   
8        2006-09-28  District of Columbia               0       Debut   
9        2006-09-28               Florida               6       Debut   

  album_id album_release_date album_release_day  periods_from_release  \
0    DEBUT         2006-10-24           Tuesday                    -2   
1    DEBUT         2006-10-24           Tuesday                    -2   
2    DEBUT         2006-10-24           Tuesday                    -2   
3    DEBUT         2006-10-24           Tuesday                    -2   
4    DEBUT         2006-10-24           Tuesday                    -2   
5    DEBUT         2006-10-24           Tuesday                    -2   
6    DEBUT         2006-10-24           Tuesday                    -2   
7    DEBUT         2006-10-24           Tuesday                    -2   
8    DEBUT         2006-10-24           Tuesday                    -2   
9    DEBUT         2006-10-24           Tuesday                    -2   

   us_search_avg us_peak_search  global_search_avg global_peak_search  
0             69     2006-10-08                 69         2006-10-07  
1             69     2006-10-08                 69         2006-10-07  
2             69     2006-10-08                 69         2006-10-07  
3             69     2006-10-08                 69         2006-10-07  
4             69     2006-10-08                 69         2006-10-07  
5             69     2006-10-08                 69         2006-10-07  
6             69     2006-10-08                 69         2006-10-07  
7             69     2006-10-08                 69         2006-10-07  
8             69     2006-10-08                 69         2006-10-07  
9             69     2006-10-08                 69         2006-10-07

In [7]:
# Check dataframe foot

df_gsearch_st.tail(10)

period_state                 period  \
3662   2023/11/08-2023/11/21, South Dakota  2023/11/08-2023/11/21   
3663      2023/11/08-2023/11/21, Tennessee  2023/11/08-2023/11/21   
3664          2023/11/08-2023/11/21, Texas  2023/11/08-2023/11/21   
3665           2023/11/08-2023/11/21, Utah  2023/11/08-2023/11/21   
3666        2023/11/08-2023/11/21, Vermont  2023/11/08-2023/11/21   
3667       2023/11/08-2023/11/21, Virginia  2023/11/08-2023/11/21   
3668     2023/11/08-2023/11/21, Washington  2023/11/08-2023/11/21   
3669  2023/11/08-2023/11/21, West Virginia  2023/11/08-2023/11/21   
3670      2023/11/08-2023/11/21, Wisconsin  2023/11/08-2023/11/21   
3671        2023/11/08-2023/11/21, Wyoming  2023/11/08-2023/11/21   

     period_start_date          state  g_search_score  \
3662        2023-11-09   South Dakota              68   
3663        2023-11-09      Tennessee              53   
3664        2023-11-09          Texas              52   
3665        2023-11-09           Utah              78   
3666        2023-11-09        Vermont              77   
3667        2023-11-09       Virginia              59   
3668        2023-11-09     Washington              63   
3669        2023-11-09  West Virginia              63   
3670        2023-11-09      Wisconsin              56   
3671        2023-11-09        Wyoming              58   

                  album_event album_id album_release_date album_release_day  \
3662  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3663  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3664  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3665  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3666  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3667  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3668  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3669  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3670  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   
3671  1989 (Taylor's Version)   1989TV         2023-10-27            Friday   

      periods_from_release  us_search_avg us_peak_search  global_search_avg  \
3662                     1             69     2023-11-12                 69   
3663                     1             69     2023-11-12                 69   
3664                     1             69     2023-11-12                 69   
3665                     1             69     2023-11-12                 69   
3666                     1             69     2023-11-12                 69   
3667                     1             69     2023-11-12                 69   
3668                     1             69     2023-11-12                 69   
3669                     1             69     2023-11-12                 69   
3670                     1             69     2023-11-12                 69   
3671                     1             69     2023-11-12                 69   

     global_peak_search  
3662         2023-11-18  
3663         2023-11-18  
3664         2023-11-18  
3665         2023-11-18  
3666         2023-11-18  
3667         2023-11-18  
3668         2023-11-18  
3669         2023-11-18  
3670         2023-11-18  
3671         2023-11-18

In [8]:
df_gsearch_st.shape

(3672, 14)

In [9]:
# Check columns, data types, and counts 

df_gsearch_st.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3672 entries, 0 to 3671
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   period_state          3672 non-null   object
 1   period                3672 non-null   object
 2   period_start_date     3672 non-null   object
 3   state                 3672 non-null   object
 4   g_search_score        3672 non-null   int64 
 5   album_event           3672 non-null   object
 6   album_id              3672 non-null   object
 7   album_release_date    3672 non-null   object
 8   album_release_day     3672 non-null   object
 9   periods_from_release  3672 non-null   int64 
 10  us_search_avg         3672 non-null   int64 
 11  us_peak_search        3672 non-null   object
 12  global_search_avg     3672 non-null   int64 
 13  global_peak_search    3672 non-null   object
dtypes: int64(4), object(10)
memory usage: 401.8+ KB


Observations: no "null" values. All data types appear correct. At this point all columns are relevant to analysis, therfore none dropped. All column names are formatted and make sense.

In [10]:
df_gsearch_st.describe()

g_search_score  periods_from_release  us_search_avg  global_search_avg
count     3672.000000           3672.000000    3672.000000        3672.000000
mean        54.990741             -0.194444      62.222222          65.222222
std         20.353841              1.595687      16.467038          15.760056
min          0.000000             -5.000000      20.000000          27.000000
25%         44.000000             -1.000000      51.750000          56.000000
50%         56.000000              0.000000      66.000000          69.500000
75%         67.000000              1.000000      75.000000          77.250000
max        100.000000              2.000000      85.000000          86.000000

Observations: Standard deviation of search scores will be a good measure for popularity in the future since the values are normalized within each period (vs. just averages). The mean of the periods_from_release is less than 0 because of the extra periods for the Midnights album. In general, state google search averages are less than US and worldwide averages, which are similar to each other. This lends itself to investigate regional trends. Since the us and global search scores are averages, they do not have a max of 100.

## Value counts of categorical variables

In [11]:
df_gsearch_st['period'].value_counts(dropna=False)

2006/09/28-2006/10/10    51
2006/10/11-2006/10/23    51
2019/08/10-2019/08/22    51
2019/07/28-2019/08/09    51
2020/08/19-2020/08/31    51
                         ..
2022/10/01-2022/10/13    51
2022/09/21-2022/10/03    51
2022/09/08-2022/09/20    51
2022/08/26-2022/09/07    51
2023/11/08-2023/11/21    51
Name: period, Length: 72, dtype: int64

In [12]:
df_gsearch_st['state'].value_counts(dropna=False)

Alabama                 72
Pennsylvania            72
Nevada                  72
New Hampshire           72
New Jersey              72
New Mexico              72
New York                72
North Carolina          72
North Dakota            72
Ohio                    72
Oklahoma                72
Oregon                  72
Rhode Island            72
Montana                 72
South Carolina          72
South Dakota            72
Tennessee               72
Texas                   72
Utah                    72
Vermont                 72
Virginia                72
Washington              72
West Virginia           72
Wisconsin               72
Nebraska                72
Missouri                72
Alaska                  72
Idaho                   72
Arizona                 72
Arkansas                72
California              72
Colorado                72
Connecticut             72
Delaware                72
District of Columbia    72
Florida                 72
Georgia                 72
H

Observations: All states have the accurate value count of 72, which matches the number of periods in the df.

In [13]:
df_gsearch_st['album_event'].value_counts(dropna=False)

Debut                           255
Speak Now                       255
1989                            255
Reputation                      255
Midnights                       255
Speak Now (Taylor's Version)    255
evermore                        255
Fearless(Taylor's Version)      255
folklore                        255
Lover                           255
Red(Taylor's Version)           255
Red                             255
Fearless                        255
1989 (Taylor's Version)         204
Midnights_announce              153
Name: album_event, dtype: int64

Observations: All value counts for album event are correct. Midnight's announcement is shorter on purpose, and 1989 (Taylor's Version) is shorter due to when the analysis started. 
-Potential need to remove space before (Taylor's Version) for Speak Now and 1989.

In [14]:
df_gsearch_st['album_id'].value_counts(dropna=False)

MIDN       408
DEBUT      255
SPEAK      255
1989       255
REP        255
SPEAKTV    255
EVER       255
FEARTV     255
FOLK       255
LOVE       255
REDTV      255
RED        255
FEAR       255
1989TV     204
Name: album_id, dtype: int64

Observations: All value counts for album_id are correct. Midnight's announcement is shorter on purpose, and 1989 (Taylor's Version) is shorter due to when the analysis started.

In [15]:
df_gsearch_st['album_release_date'].value_counts(dropna=False)

2022-10-27    408
2006-10-24    255
2010-10-25    255
2014-10-27    255
2017-11-10    255
2023-07-07    255
2020-12-11    255
2021-04-09    255
2020-07-24    255
2019-08-23    255
2021-11-12    255
2012-10-22    255
2008-11-11    255
2023-10-27    204
Name: album_release_date, dtype: int64

Observation: All dates accurate and formatted consistently.

In [16]:
df_gsearch_st['album_release_day'].value_counts(dropna=False)

Friday      1989
Monday       765
Tuesday      510
Thursday     408
Name: album_release_day, dtype: int64

Observation: Taylor has never released an album on Wednesday, Saturday, or Sunday. 

In [17]:
df_gsearch_st['periods_from_release'].value_counts(dropna=False)

-2    714
-1    714
 0    714
 1    714
 2    663
-5     51
-4     51
-3     51
Name: periods_from_release, dtype: int64

Observation: Period counts match due to reasons above, with Midnights event adding -3, -4, and -5 and 1989(Taylor's Version) only having weeks -2 through 1. All other albums have periods -2, -1, 0, 1, and 2

## Check mixed data types

In [18]:
for col in df_gsearch_st.columns.tolist():
    weird = (df_gsearch_st[[col]].applymap(type) !=df_gsearch_st[[col]].iloc[0].apply(type)).any(axis = 1)                                                                             
    if len (df_gsearch_st[weird]) > 0:
        print (col)

Observations: no mixed type columns, error stems from search all columns with "apply" instead of map with specific columns

## Check missing values

In [19]:
df_gsearch_st.isnull().sum()

period_state            0
period                  0
period_start_date       0
state                   0
g_search_score          0
album_event             0
album_id                0
album_release_date      0
album_release_day       0
periods_from_release    0
us_search_avg           0
us_peak_search          0
global_search_avg       0
global_peak_search      0
dtype: int64

## Check for duplicates

In [20]:
df_gsearch_st_dupes = df_gsearch_st[df_gsearch_st.duplicated()]

In [21]:
df_gsearch_st_dupes

Empty DataFrame
Columns: [period_state, period, period_start_date, state, g_search_score, album_event, album_id, album_release_date, album_release_day, periods_from_release, us_search_avg, us_peak_search, global_search_avg, global_peak_search]
Index: []

Observations: no missing values and no complete duplicates. Data frame is cleaned and ready for export

## Final cleaned summary stats and variables

In [22]:
df_gsearch_st.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3672 entries, 0 to 3671
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   period_state          3672 non-null   object
 1   period                3672 non-null   object
 2   period_start_date     3672 non-null   object
 3   state                 3672 non-null   object
 4   g_search_score        3672 non-null   int64 
 5   album_event           3672 non-null   object
 6   album_id              3672 non-null   object
 7   album_release_date    3672 non-null   object
 8   album_release_day     3672 non-null   object
 9   periods_from_release  3672 non-null   int64 
 10  us_search_avg         3672 non-null   int64 
 11  us_peak_search        3672 non-null   object
 12  global_search_avg     3672 non-null   int64 
 13  global_peak_search    3672 non-null   object
dtypes: int64(4), object(10)
memory usage: 401.8+ KB


In [23]:
df_gsearch_st.shape

(3672, 14)

In [24]:
df_gsearch_st.describe()

g_search_score  periods_from_release  us_search_avg  global_search_avg
count     3672.000000           3672.000000    3672.000000        3672.000000
mean        54.990741             -0.194444      62.222222          65.222222
std         20.353841              1.595687      16.467038          15.760056
min          0.000000             -5.000000      20.000000          27.000000
25%         44.000000             -1.000000      51.750000          56.000000
50%         56.000000              0.000000      66.000000          69.500000
75%         67.000000              1.000000      75.000000          77.250000
max        100.000000              2.000000      85.000000          86.000000

Cleaned dataframe exported as gsearch_st_clean

# 03. song_info data cleaning and wrangling

In [25]:
# Check dataframe import

df_song_info.head()

Unnamed: 0                                           name  \
0           0         Welcome To New York (Taylor's Version)   
1           1                 Blank Space (Taylor's Version)   
2           2                       Style (Taylor's Version)   
3           3            Out Of The Woods (Taylor's Version)   
4           4  All You Had To Do Was Stay (Taylor's Version)   

                              album release_date  track_number  \
0  1989 (Taylor's Version) [Deluxe]   10/27/2023             1   
1  1989 (Taylor's Version) [Deluxe]   10/27/2023             2   
2  1989 (Taylor's Version) [Deluxe]   10/27/2023             3   
3  1989 (Taylor's Version) [Deluxe]   10/27/2023             4   
4  1989 (Taylor's Version) [Deluxe]   10/27/2023             5   

                       id                                   uri  acousticness  \
0  4WUepByoeqcedHoYhSNHRt  spotify:track:4WUepByoeqcedHoYhSNHRt      0.009420   
1  0108kcWLnn2HlH2kedi1gn  spotify:track:0108kcWLnn2HlH2kedi1gn      0.088500   
2  3Vpk1hfMAQme8VJ0SNRSkd  spotify:track:3Vpk1hfMAQme8VJ0SNRSkd      0.000421   
3  1OcSfkeCg9hRC2sFKB4IMJ  spotify:track:1OcSfkeCg9hRC2sFKB4IMJ      0.000537   
4  2k0ZEeAqzvYMcx9Qt5aClQ  spotify:track:2k0ZEeAqzvYMcx9Qt5aClQ      0.000656   

   danceability  energy  instrumentalness  liveness  loudness  speechiness  \
0         0.757   0.610          0.000037    0.3670    -4.840       0.0327   
1         0.733   0.733          0.000000    0.1680    -5.376       0.0670   
2         0.511   0.822          0.019700    0.0899    -4.785       0.0397   
3         0.545   0.885          0.000056    0.3850    -5.968       0.0447   
4         0.588   0.721          0.000000    0.1310    -5.579       0.0317   

     tempo  valence  popularity  duration_ms  
0  116.998    0.685          80       212600  
1   96.057    0.701          80       231833  
2   94.868    0.305          81       231000  
3   92.021    0.206          80       235800  
4   96.997    0.520          79       193289

In [26]:
df_song_info.tail()

Unnamed: 0                                  name         album  \
525         525                              Our Song  Taylor Swift   
526         526         I'm Only Me When I'm With You  Taylor Swift   
527         527                             Invisible  Taylor Swift   
528         528                A Perfectly Good Heart  Taylor Swift   
529         529  Teardrops on My Guitar - Pop Version  Taylor Swift   

    release_date  track_number                      id  \
525   10/24/2006            11  15DeqWWQB4dcEWzJg15VrN   
526   10/24/2006            12  0JIdBrXGSJXS72zjF9ss9u   
527   10/24/2006            13  5OOd01o2YS1QFwdpVLds3r   
528   10/24/2006            14  1spLfUJxtyVyiKKTegQ2r4   
529   10/24/2006            15  4pJi1rVt9GNegU9kywjg4z   

                                      uri  acousticness  danceability  energy  \
525  spotify:track:15DeqWWQB4dcEWzJg15VrN       0.11100         0.668   0.672   
526  spotify:track:0JIdBrXGSJXS72zjF9ss9u       0.00452         0.563   0.934   
527  spotify:track:5OOd01o2YS1QFwdpVLds3r       0.63700         0.612   0.394   
528  spotify:track:1spLfUJxtyVyiKKTegQ2r4       0.00349         0.483   0.751   
529  spotify:track:4pJi1rVt9GNegU9kywjg4z       0.04020         0.459   0.753   

     instrumentalness  liveness  loudness  speechiness    tempo  valence  \
525          0.000000    0.3290    -4.931       0.0303   89.011    0.539   
526          0.000807    0.1030    -3.629       0.0646  143.964    0.518   
527          0.000000    0.1470    -5.723       0.0243   96.001    0.233   
528          0.000000    0.1280    -5.726       0.0365  156.092    0.268   
529          0.000000    0.0863    -3.827       0.0537  199.997    0.483   

     popularity  duration_ms  
525          77       201106  
526          62       213053  
527          58       203226  
528          57       220146  
529          57       179066

In [27]:
df_song_info.shape

(530, 18)

In [28]:
df_song_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 530 entries, 0 to 529
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        530 non-null    int64  
 1   name              530 non-null    object 
 2   album             530 non-null    object 
 3   release_date      530 non-null    object 
 4   track_number      530 non-null    int64  
 5   id                530 non-null    object 
 6   uri               530 non-null    object 
 7   acousticness      530 non-null    float64
 8   danceability      530 non-null    float64
 9   energy            530 non-null    float64
 10  instrumentalness  530 non-null    float64
 11  liveness          530 non-null    float64
 12  loudness          530 non-null    float64
 13  speechiness       530 non-null    float64
 14  tempo             530 non-null    float64
 15  valence           530 non-null    float64
 16  popularity        530 non-null    int64  
 1

Observations: 
-All columns have accurate 530 rows.
-Need to drop "Unnamed: 0"
-Can drop id and uri colums
-Need to update name, album, release_date, popularity, and duration_ms to be more clear and match other dataframes
-Need to change track_number to object string, but all other data types appear accurate
-For cohesion, will add an album_id and song_id to match supplemental dataframes and prepare for future analysis steps
-Can drop all rows of songs outside of albums covered in this analysis, will save as separate df

## Column changes

In [29]:
df_song_info.drop(columns=['Unnamed: 0', 'id', 'uri'], axis=1, inplace=True)

In [30]:
df_song_info.columns

Index(['name', 'album', 'release_date', 'track_number', 'acousticness',
       'danceability', 'energy', 'instrumentalness', 'liveness', 'loudness',
       'speechiness', 'tempo', 'valence', 'popularity', 'duration_ms'],
      dtype='object')

In [31]:
df_song_info.rename(columns = {'name': 'song_name', 'release_date': 'album_release_date', 'popularity': 'song_popularity', 'duration': 'song_length'}, inplace=True)

In [32]:
# Check rename
df_song_info.columns

Index(['song_name', 'album', 'album_release_date', 'track_number',
       'acousticness', 'danceability', 'energy', 'instrumentalness',
       'liveness', 'loudness', 'speechiness', 'tempo', 'valence',
       'song_popularity', 'duration_ms'],
      dtype='object')

## Data types

Change track_number to string object, will create a new column for stat potential for both integers and strings manipulation

In [33]:
# Add column for track_number as string

df_song_info['track_number_str'] = df_song_info['track_number']

In [34]:
df_song_info.head(5)

song_name  \
0         Welcome To New York (Taylor's Version)   
1                 Blank Space (Taylor's Version)   
2                       Style (Taylor's Version)   
3            Out Of The Woods (Taylor's Version)   
4  All You Had To Do Was Stay (Taylor's Version)   

                              album album_release_date  track_number  \
0  1989 (Taylor's Version) [Deluxe]         10/27/2023             1   
1  1989 (Taylor's Version) [Deluxe]         10/27/2023             2   
2  1989 (Taylor's Version) [Deluxe]         10/27/2023             3   
3  1989 (Taylor's Version) [Deluxe]         10/27/2023             4   
4  1989 (Taylor's Version) [Deluxe]         10/27/2023             5   

   acousticness  danceability  energy  instrumentalness  liveness  loudness  \
0      0.009420         0.757   0.610          0.000037    0.3670    -4.840   
1      0.088500         0.733   0.733          0.000000    0.1680    -5.376   
2      0.000421         0.511   0.822          0.019700    0.0899    -4.785   
3      0.000537         0.545   0.885          0.000056    0.3850    -5.968   
4      0.000656         0.588   0.721          0.000000    0.1310    -5.579   

   speechiness    tempo  valence  song_popularity  duration_ms  \
0       0.0327  116.998    0.685               80       212600   
1       0.0670   96.057    0.701               80       231833   
2       0.0397   94.868    0.305               81       231000   
3       0.0447   92.021    0.206               80       235800   
4       0.0317   96.997    0.520               79       193289   

   track_number_str  
0                 1  
1                 2  
2                 3  
3                 4  
4                 5

In [35]:
# Change new column to a string
df_song_info['track_number_str'] = df_song_info['track_number_str'].astype(str)

In [36]:
# Check type

df_song_info['track_number_str'].dtype

dtype('O')

In [37]:
df_song_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 530 entries, 0 to 529
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   song_name           530 non-null    object 
 1   album               530 non-null    object 
 2   album_release_date  530 non-null    object 
 3   track_number        530 non-null    int64  
 4   acousticness        530 non-null    float64
 5   danceability        530 non-null    float64
 6   energy              530 non-null    float64
 7   instrumentalness    530 non-null    float64
 8   liveness            530 non-null    float64
 9   loudness            530 non-null    float64
 10  speechiness         530 non-null    float64
 11  tempo               530 non-null    float64
 12  valence             530 non-null    float64
 13  song_popularity     530 non-null    int64  
 14  duration_ms         530 non-null    int64  
 15  track_number_str    530 non-null    object 
dtypes: float

In [38]:
# Initial descriptive stats

df_song_info.describe()

track_number  acousticness  danceability      energy  instrumentalness  \
count    530.000000    530.000000    530.000000  530.000000        530.000000   
mean      11.184906      0.319247      0.585285    0.574609          0.004005   
std        7.981060      0.327043      0.113121    0.191565          0.033163   
min        1.000000      0.000184      0.243000    0.118000          0.000000   
25%        5.000000      0.036250      0.516000    0.443000          0.000000   
50%       10.000000      0.165000      0.595500    0.589500          0.000002   
75%       15.000000      0.653000      0.653000    0.729750          0.000056   
max       46.000000      0.971000      0.897000    0.950000          0.488000   

         liveness    loudness  speechiness       tempo     valence  \
count  530.000000  530.000000   530.000000  530.000000  530.000000   
mean     0.163492   -7.505434     0.055889  122.332311    0.397379   
std      0.142263    2.939040     0.070300   30.000272    0.199589   
min      0.035700  -17.932000     0.023100   68.097000    0.037400   
25%      0.096600   -9.222000     0.030325   96.939500    0.230000   
50%      0.115000   -7.012000     0.037300  119.030000    0.385500   
75%      0.163000   -5.361750     0.055225  143.934250    0.535000   
max      0.931000   -1.909000     0.912000  208.918000    0.943000   

       song_popularity    duration_ms  
count       530.000000     530.000000  
mean         64.852830  239978.624528  
std          15.414532   46119.983031  
min          33.000000   83253.000000  
25%          52.000000  211813.000000  
50%          69.000000  235273.000000  
75%          77.000000  260361.000000  
max         100.000000  613026.000000

Observations: 
-Need to investigate track_number max of 46, seems to high
-Want to investigate loudness scores, since max is negative and mean is positive.
-At this point all other figures appear accurate

## Value counts

In [39]:
df_song_info['song_name'].value_counts()

Question...?                                         3
Anti-Hero                                            3
Labyrinth                                            3
Bejeweled                                            3
Vigilante Shit                                       3
                                                    ..
my tears ricochet - the long pond studio sessions    1
mirrorball - the long pond studio sessions           1
seven - the long pond studio sessions                1
this is me trying - the long pond studio sessions    1
Teardrops on My Guitar - Pop Version                 1
Name: song_name, Length: 329, dtype: int64

Note: there are duplicate song names for all values > 1. Will address more duplicates after I remove unneeded albums

In [40]:
df_song_info['album'].value_counts()

reputation Stadium Tour Surprise Song Playlist                                         46
folklore: the long pond studio sessions (from the Disney+ special) [deluxe edition]    34
Red (Taylor's Version)                                                                 30
Fearless (Taylor's Version)                                                            26
Midnights (The Til Dawn Edition)                                                       23
1989 (Taylor's Version) [Deluxe]                                                       22
Speak Now (Taylor's Version)                                                           22
Red (Deluxe Edition)                                                                   22
1989 (Taylor's Version)                                                                21
Midnights (3am Edition)                                                                20
Speak Now (Deluxe Edition)                                                             20
Fearless P

In [41]:
df_song_info['track_number'].value_counts().sort_values(ascending=False)

1     28
5     28
4     28
3     28
6     28
2     28
8     28
7     28
13    27
10    27
9     27
12    27
11    27
14    24
15    23
16    20
17    17
18    13
19    12
20    10
21     8
22     7
23     4
25     3
26     3
24     3
29     2
30     2
27     2
28     2
38     1
31     1
32     1
33     1
34     1
35     1
37     1
41     1
36     1
40     1
42     1
43     1
44     1
45     1
39     1
46     1
Name: track_number, dtype: int64

Note: will need to address formatting so that tracks are in order prior to song id. Will use track_number_str for this

In [42]:
# Check value counts for Reputation Surprise song list, and see if that accounts for all higher counts above

df_high_song_count = df_song_info[df_song_info['track_number'] >20]

In [43]:
df_high_song_count

song_name  \
20   Is It Over Now? (Taylor's Version) (From The V...   
21   Bad Blood (feat. Kendrick Lamar) (Taylor's Ver...   
42   Is It Over Now? (Taylor's Version) (From The V...   
63     Foolish One (Taylor's Version) (From The Vault)   
64        Timeless (Taylor's Version) (From The Vault)   
85                                      Hits Different   
86         Snow On The Beach (feat. More Lana Del Rey)   
87                             Karma (feat. Ice Spice)   
141                           Ronan (Taylor's Version)   
142     Better Man (Taylor's Version) (From The Vault)   
143  Nothing New (feat. Phoebe Bridgers) (Taylor's ...   
144           Babe (Taylor's Version) (From The Vault)   
145  Message In A Bottle (Taylor's Version) (From T...   
146  I Bet You Think About Me (feat. Chris Stapleto...   
147  Forever Winter (Taylor's Version) (From The Va...   
148  Run (feat. Ed Sheeran) (Taylor's Version) (Fro...   
149  The Very First Night (Taylor's Version) (From ...   
150  All Too Well (10 Minute Version) (Taylor's Ver...   
171  You All Over Me (feat. Maren Morris) (Taylor's...   
172  Mr. Perfectly Fine (Taylor's Version) (From Th...   
173  We Were Happy (Taylor's Version) (From The Vault)   
174  That's When (feat. Keith Urban) (Taylor's Vers...   
175      Don't You (Taylor's Version) (From The Vault)   
176   Bye Bye Baby (Taylor's Version) (From The Vault)   
329                                               Babe   
330                                Welcome To New York   
331                                           Fearless   
332                                          Enchanted   
333                                             Change   
334                                               Ours   
335                                   Out Of The Woods   
336                                Come Back...Be Here   
337                              A Place in this World   
338                                          This Love   
339                                      The Lucky One   
340                                          Invisible   
341                                            Breathe   
342                                         Better Man   
343                                     Jump Then Fall   
344                                        Begin Again   
345                         Tied Together with a Smile   
346                                    The Story Of Us   
347                                   Forever & Always   
348                                        Hey Stephen   
349                                          Speak Now   
350                                         Wonderland   
351                                        White Horse   
352                      I'm Only Me When I'm With You   
353                                          Starlight   
354                                      I Know Places   
407                      Red - Original Demo Recording   
408                          State Of Grace - Acoustic   

                                              album album_release_date  \
20                 1989 (Taylor's Version) [Deluxe]         10/27/2023   
21                 1989 (Taylor's Version) [Deluxe]         10/27/2023   
42                          1989 (Taylor's Version)         10/26/2023   
63                     Speak Now (Taylor's Version)           7/7/2023   
64                     Speak Now (Taylor's Version)           7/7/2023   
85                 Midnights (The Til Dawn Edition)          5/26/2023   
86                 Midnights (The Til Dawn Edition)          5/26/2023   
87                 Midnights (The Til Dawn Edition)          5/26/2023   
141                          Red (Taylor's Version)         11/12/2021   
142                          Red (Taylor's Version)         11/12/2021   
143                          Red (Taylor's Version)         11/12/2021   
144                          Red (Taylor's Version)         11/12/2021   
145     

Observations: all track_numbers are accurate, and high counts come from expectedly long Rep Stadium tour. Once remove unneeded albums will recalcuate stats.

## Remove unneeded albums

In [44]:
df_song_info['album'].value_counts(dropna=False)

reputation Stadium Tour Surprise Song Playlist                                         46
folklore: the long pond studio sessions (from the Disney+ special) [deluxe edition]    34
Red (Taylor's Version)                                                                 30
Fearless (Taylor's Version)                                                            26
Midnights (The Til Dawn Edition)                                                       23
1989 (Taylor's Version) [Deluxe]                                                       22
Speak Now (Taylor's Version)                                                           22
Red (Deluxe Edition)                                                                   22
1989 (Taylor's Version)                                                                21
Midnights (3am Edition)                                                                20
Speak Now (Deluxe Edition)                                                             20
Fearless P

In [45]:
df_song_info.shape

(530, 16)

In [99]:
# Select dataframe to use with only relevant albums. Choosing the album with the most songs (ex: Deluxe) to include complete relevant song list.

df_song_info_stud_alb = df_song_info.loc[df_song_info['album'].isin(['Taylor Swift',
                                                                    'Fearless Platinum Edition',
                                                                    'Speak Now (Deluxe Edition)',
                                                                    'Red (Deluxe Edition)',
                                                                    '1989 (Deluxe Edition)',
                                                                    'reputation',
                                                                    'Lover',
                                                                    'folklore (deluxe version)',
                                                                    'evermore (deluxe version)',
                                                                    "Fearless (Taylor's Version)",
                                                                    "Red (Taylor's Version)",
                                                                    "Midnights (The Til Dawn Edition)",
                                                                    "Speak Now (Taylor's Version)",
                                                                    "1989 (Taylor's Version) [Deluxe]"])]

                                                     
                                                     
                                                     
                                                     
                                                    
                                                     
                                                     

In [100]:
df_song_info_stud_alb.shape

(285, 16)

In [101]:
df_song_info_stud_alb.head()

song_name  \
0         Welcome To New York (Taylor's Version)   
1                 Blank Space (Taylor's Version)   
2                       Style (Taylor's Version)   
3            Out Of The Woods (Taylor's Version)   
4  All You Had To Do Was Stay (Taylor's Version)   

                              album album_release_date  track_number  \
0  1989 (Taylor's Version) [Deluxe]         10/27/2023             1   
1  1989 (Taylor's Version) [Deluxe]         10/27/2023             2   
2  1989 (Taylor's Version) [Deluxe]         10/27/2023             3   
3  1989 (Taylor's Version) [Deluxe]         10/27/2023             4   
4  1989 (Taylor's Version) [Deluxe]         10/27/2023             5   

   acousticness  danceability  energy  instrumentalness  liveness  loudness  \
0      0.009420         0.757   0.610          0.000037    0.3670    -4.840   
1      0.088500         0.733   0.733          0.000000    0.1680    -5.376   
2      0.000421         0.511   0.822          0.019700    0.0899    -4.785   
3      0.000537         0.545   0.885          0.000056    0.3850    -5.968   
4      0.000656         0.588   0.721          0.000000    0.1310    -5.579   

   speechiness    tempo  valence  song_popularity  duration_ms  \
0       0.0327  116.998    0.685               80       212600   
1       0.0670   96.057    0.701               80       231833   
2       0.0397   94.868    0.305               81       231000   
3       0.0447   92.021    0.206               80       235800   
4       0.0317   96.997    0.520               79       193289   

  track_number_str  
0                1  
1                2  
2                3  
3                4  
4                5

In [102]:
df_song_info_stud_alb['album'].value_counts(dropna=False)

Red (Taylor's Version)              30
Fearless (Taylor's Version)         26
Midnights (The Til Dawn Edition)    23
1989 (Taylor's Version) [Deluxe]    22
Speak Now (Taylor's Version)        22
Red (Deluxe Edition)                22
Speak Now (Deluxe Edition)          20
1989 (Deluxe Edition)               19
Fearless Platinum Edition           19
Lover                               18
evermore (deluxe version)           17
folklore (deluxe version)           17
reputation                          15
Taylor Swift                        15
Name: album, dtype: int64

In [103]:
# Subset dataframe for unused albums

df_song_info_no_use_albums = df_song_info.loc[df_song_info['album'].isin(["reputation Stadium Tour Surprise Song Playlist",
                                "folklore: the long pond studio sessions (from the Disney+ special) [deluxe edition]",
                                "1989 (Taylor's Version)",
                                "Midnights (3am Edition)",
                                "Folklore",
                                "Red",
                                "Speak Now World Tour Live",
                                "evermore",
                                "Speak Now",
                                "1989",
                                "Midnights",
                                "Fearless",
                                "Live From Clear Channel Stripped 2008"])]

In [104]:
df_song_info_no_use_albums.shape

(229, 16)

In [105]:
df_song_info_no_use_albums.head()

song_name                    album  \
22         Welcome To New York (Taylor's Version)  1989 (Taylor's Version)   
23                 Blank Space (Taylor's Version)  1989 (Taylor's Version)   
24                       Style (Taylor's Version)  1989 (Taylor's Version)   
25            Out Of The Woods (Taylor's Version)  1989 (Taylor's Version)   
26  All You Had To Do Was Stay (Taylor's Version)  1989 (Taylor's Version)   

   album_release_date  track_number  acousticness  danceability  energy  \
22         10/26/2023             1      0.009610         0.761   0.607   
23         10/26/2023             2      0.097600         0.737   0.725   
24         10/26/2023             3      0.000418         0.575   0.824   
25         10/26/2023             4      0.000512         0.547   0.885   
26         10/26/2023             5      0.000718         0.586   0.720   

    instrumentalness  liveness  loudness  speechiness    tempo  valence  \
22          0.000022    0.3670    -4.830       0.0312  116.980    0.674   
23          0.000000    0.1740    -5.417       0.0641   96.012    0.755   
24          0.021400    0.0901    -4.823       0.0380   94.971    0.292   
25          0.000094    0.3900    -5.961       0.0418   91.992    0.192   
26          0.000000    0.0949    -5.556       0.0350   97.057    0.481   

    song_popularity  duration_ms track_number_str  
22               87       212600                1  
23               88       231833                2  
24               89       231000                3  
25               87       235800                4  
26               86       193289                5

In [106]:
df_song_info_no_use_albums['album'].value_counts(dropna=False)

reputation Stadium Tour Surprise Song Playlist                                         46
folklore: the long pond studio sessions (from the Disney+ special) [deluxe edition]    34
1989 (Taylor's Version)                                                                21
Midnights (3am Edition)                                                                20
Red                                                                                    16
Speak Now World Tour Live                                                              16
evermore                                                                               15
Speak Now                                                                              14
Midnights                                                                              13
1989                                                                                   13
Fearless                                                                               13
Live From 

In [107]:
df_song_info_stud_alb['album'].value_counts(dropna=False)

Red (Taylor's Version)              30
Fearless (Taylor's Version)         26
Midnights (The Til Dawn Edition)    23
1989 (Taylor's Version) [Deluxe]    22
Speak Now (Taylor's Version)        22
Red (Deluxe Edition)                22
Speak Now (Deluxe Edition)          20
1989 (Deluxe Edition)               19
Fearless Platinum Edition           19
Lover                               18
evermore (deluxe version)           17
folklore (deluxe version)           17
reputation                          15
Taylor Swift                        15
Name: album, dtype: int64

Observations: Successfully removed unneeded albums. Counts match track number counts. Dataframe exported at end of script

In [108]:
# Check song counts to all be 1 

df_song_info_stud_alb['song_name'].value_counts()

Welcome To New York (Taylor's Version)               1
Call It What You Want                                1
All You Had To Do Was Stay                           1
Out Of The Woods                                     1
Style                                                1
                                                    ..
White Horse (Taylor's Version)                       1
You Belong With Me (Taylor's Version)                1
Breathe (feat. Colbie Caillat) (Taylor's Version)    1
Tell Me Why (Taylor's Version)                       1
Teardrops on My Guitar - Pop Version                 1
Name: song_name, Length: 285, dtype: int64

## Check for duplicates

In [109]:
df_song_info_stud_alb_dupes = df_song_info_stud_alb[df_song_info_stud_alb.duplicated()]

In [110]:
df_song_info_stud_alb_dupes

Empty DataFrame
Columns: [song_name, album, album_release_date, track_number, acousticness, danceability, energy, instrumentalness, liveness, loudness, speechiness, tempo, valence, song_popularity, duration_ms, track_number_str]
Index: []

No full duplicates found.

## Check for missing values

In [111]:
df_song_info_stud_alb.isnull().sum()

song_name             0
album                 0
album_release_date    0
track_number          0
acousticness          0
danceability          0
energy                0
instrumentalness      0
liveness              0
loudness              0
speechiness           0
tempo                 0
valence               0
song_popularity       0
duration_ms           0
track_number_str      0
dtype: int64

No missing values.

## Check track_number formatting

In [112]:
df_song_info_stud_alb['track_number'].value_counts(ascending =False)

1     14
3     14
4     14
5     14
6     14
7     14
8     14
9     14
10    14
11    14
12    14
13    14
14    14
15    14
2     14
17    12
16    12
18    10
19     9
20     7
21     6
22     6
23     3
24     2
25     2
26     2
27     1
28     1
29     1
30     1
Name: track_number, dtype: int64

In [113]:
# Use "track_number_str" column to create a 2 digit value for all track numbers, which will be used to have proper order of songs and build song ids

df_song_info_stud_alb['track_number_str'] = df_song_info_stud_alb['track_number_str'].astype(str).str.pad(2, side='left', fillchar='0')

C:\Users\leobs\AppData\Local\Temp\ipykernel_14464\3953563745.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_song_info_stud_alb['track_number_str'] = df_song_info_stud_alb['track_number_str'].astype(str).str.pad(2, side='left', fillchar='0')


In [114]:
df_song_info_stud_alb['track_number'].value_counts(ascending =False)

1     14
3     14
4     14
5     14
6     14
7     14
8     14
9     14
10    14
11    14
12    14
13    14
14    14
15    14
2     14
17    12
16    12
18    10
19     9
20     7
21     6
22     6
23     3
24     2
25     2
26     2
27     1
28     1
29     1
30     1
Name: track_number, dtype: int64

In [115]:
df_song_info_stud_alb["track_number_str"].replace({'0':'01'},inplace=True)

C:\Users\leobs\AppData\Local\Temp\ipykernel_14464\3140419998.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_song_info_stud_alb["track_number_str"].replace({'0':'01'},inplace=True)


In [116]:
df_song_info_stud_alb.head()

song_name  \
0         Welcome To New York (Taylor's Version)   
1                 Blank Space (Taylor's Version)   
2                       Style (Taylor's Version)   
3            Out Of The Woods (Taylor's Version)   
4  All You Had To Do Was Stay (Taylor's Version)   

                              album album_release_date  track_number  \
0  1989 (Taylor's Version) [Deluxe]         10/27/2023             1   
1  1989 (Taylor's Version) [Deluxe]         10/27/2023             2   
2  1989 (Taylor's Version) [Deluxe]         10/27/2023             3   
3  1989 (Taylor's Version) [Deluxe]         10/27/2023             4   
4  1989 (Taylor's Version) [Deluxe]         10/27/2023             5   

   acousticness  danceability  energy  instrumentalness  liveness  loudness  \
0      0.009420         0.757   0.610          0.000037    0.3670    -4.840   
1      0.088500         0.733   0.733          0.000000    0.1680    -5.376   
2      0.000421         0.511   0.822          0.019700    0.0899    -4.785   
3      0.000537         0.545   0.885          0.000056    0.3850    -5.968   
4      0.000656         0.588   0.721          0.000000    0.1310    -5.579   

   speechiness    tempo  valence  song_popularity  duration_ms  \
0       0.0327  116.998    0.685               80       212600   
1       0.0670   96.057    0.701               80       231833   
2       0.0397   94.868    0.305               81       231000   
3       0.0447   92.021    0.206               80       235800   
4       0.0317   96.997    0.520               79       193289   

  track_number_str  
0               01  
1               02  
2               03  
3               04  
4               05

In [117]:
df_song_info_stud_alb.tail()

song_name         album album_release_date  \
525                              Our Song  Taylor Swift         10/24/2006   
526         I'm Only Me When I'm With You  Taylor Swift         10/24/2006   
527                             Invisible  Taylor Swift         10/24/2006   
528                A Perfectly Good Heart  Taylor Swift         10/24/2006   
529  Teardrops on My Guitar - Pop Version  Taylor Swift         10/24/2006   

     track_number  acousticness  danceability  energy  instrumentalness  \
525            11       0.11100         0.668   0.672          0.000000   
526            12       0.00452         0.563   0.934          0.000807   
527            13       0.63700         0.612   0.394          0.000000   
528            14       0.00349         0.483   0.751          0.000000   
529            15       0.04020         0.459   0.753          0.000000   

     liveness  loudness  speechiness    tempo  valence  song_popularity  \
525    0.3290    -4.931       0.0303   89.011    0.539               77   
526    0.1030    -3.629       0.0646  143.964    0.518               62   
527    0.1470    -5.723       0.0243   96.001    0.233               58   
528    0.1280    -5.726       0.0365  156.092    0.268               57   
529    0.0863    -3.827       0.0537  199.997    0.483               57   

     duration_ms track_number_str  
525       201106               11  
526       213053               12  
527       203226               13  
528       220146               14  
529       179066               15

In [118]:
df_song_info_stud_alb.columns

Index(['song_name', 'album', 'album_release_date', 'track_number',
       'acousticness', 'danceability', 'energy', 'instrumentalness',
       'liveness', 'loudness', 'speechiness', 'tempo', 'valence',
       'song_popularity', 'duration_ms', 'track_number_str'],
      dtype='object')

In [119]:
df_song_info_stud_alb['track_number_str'].value_counts(ascending =False)

01    14
03    14
04    14
05    14
06    14
07    14
08    14
09    14
10    14
11    14
12    14
13    14
14    14
15    14
02    14
17    12
16    12
18    10
19     9
20     7
21     6
22     6
23     3
24     2
25     2
26     2
27     1
28     1
29     1
30     1
Name: track_number_str, dtype: int64

Observation: Successful conversion of track_number_str to xx format. Can now use to create song_id and also can maintain numerical value of track_number to use for calcs.    

## Initial column derivatives

Goal: need to add album_id to each album and create song_id based on album_id and track_number_str

In [120]:
df_song_info_stud_alb['album'].value_counts()

Red (Taylor's Version)              30
Fearless (Taylor's Version)         26
Midnights (The Til Dawn Edition)    23
1989 (Taylor's Version) [Deluxe]    22
Speak Now (Taylor's Version)        22
Red (Deluxe Edition)                22
Speak Now (Deluxe Edition)          20
1989 (Deluxe Edition)               19
Fearless Platinum Edition           19
Lover                               18
evermore (deluxe version)           17
folklore (deluxe version)           17
reputation                          15
Taylor Swift                        15
Name: album, dtype: int64

In [121]:
# Rename Taylor Swift to Debut, for clarity and accuracy

df_song_info_stud_alb['album'].replace({"Taylor Swift":"Debut"}, inplace=True)

C:\Users\leobs\AppData\Local\Temp\ipykernel_14464\1789793212.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_song_info_stud_alb['album'].replace({"Taylor Swift":"Debut"}, inplace=True)


In [122]:
check_ts_deb = df_song_info_stud_alb[df_song_info_stud_alb['album'] == "Taylor Swift"]

In [123]:
check_ts_deb

Empty DataFrame
Columns: [song_name, album, album_release_date, track_number, acousticness, danceability, energy, instrumentalness, liveness, loudness, speechiness, tempo, valence, song_popularity, duration_ms, track_number_str]
Index: []

In [124]:
check_deb = df_song_info_stud_alb[df_song_info_stud_alb['album'] == "Debut"]

In [125]:
check_deb

song_name  album album_release_date  \
515                                   Tim McGraw  Debut         10/24/2006   
516                              Picture To Burn  Debut         10/24/2006   
517  Teardrops On My Guitar - Radio Single Remix  Debut         10/24/2006   
518                        A Place in this World  Debut         10/24/2006   
519                                  Cold As You  Debut         10/24/2006   
520                                  The Outside  Debut         10/24/2006   
521                   Tied Together with a Smile  Debut         10/24/2006   
522                               Stay Beautiful  Debut         10/24/2006   
523                            Should've Said No  Debut         10/24/2006   
524                    Mary's Song (Oh My My My)  Debut         10/24/2006   
525                                     Our Song  Debut         10/24/2006   
526                I'm Only Me When I'm With You  Debut         10/24/2006   
527                                    Invisible  Debut         10/24/2006   
528                       A Perfectly Good Heart  Debut         10/24/2006   
529         Teardrops on My Guitar - Pop Version  Debut         10/24/2006   

     track_number  acousticness  danceability  energy  instrumentalness  \
515             1       0.57500         0.580   0.491          0.000000   
516             2       0.17300         0.658   0.877          0.000000   
517             3       0.28800         0.621   0.417          0.000000   
518             4       0.05100         0.576   0.777          0.000000   
519             5       0.21700         0.418   0.482          0.000000   
520             6       0.00491         0.589   0.805          0.000000   
521             7       0.52500         0.479   0.578          0.000000   
522             8       0.08680         0.594   0.629          0.000000   
523             9       0.01030         0.476   0.777          0.000000   
524            10       0.01770         0.403   0.627          0.000000   
525            11       0.11100         0.668   0.672          0.000000   
526            12       0.00452         0.563   0.934          0.000807   
527            13       0.63700         0.612   0.394          0.000000   
528            14       0.00349         0.483   0.751          0.000000   
529            15       0.04020         0.459   0.753          0.000000   

     liveness  loudness  speechiness    tempo  valence  song_popularity  \
515    0.1210    -6.462       0.0251   76.009    0.425               68   
516    0.0962    -2.098       0.0323  105.586    0.821               72   
517    0.1190    -6.941       0.0231   99.953    0.289               69   
518    0.3200    -2.881       0.0324  115.028    0.428               60   
519    0.1230    -5.769       0.0266  175.558    0.261               60   
520    0.2400    -4.055       0.0293  112.982    0.591               58   
521    0.0841    -4.963       0.0294  146.165    0.192               59   
522    0.1370    -4.919       0.0246  131.597    0.504               59   
523    0.1960    -3.771       0.0289  167.964    0.472               69   
524    0.1820    -5.280       0.0292   74.900    0.374               62   
525    0.3290    -4.931       0.0303   89.011    0.539               77   
526    0.1030    -3.629       0.0646  143.964    0.518               62   
527    0.1470    -5.723       0.0243   96.001    0.233               58   
528    0.1280    -5.726       0.0365  156.092    0.268               57   
529    0.0863    -3.827       0.0537  199.997    0.483               57   

     duration_ms track_number_str  
515       232106               01  
516       173066               02  
517       203040               03  
518       199200               04  
519       239013               05  
520       207106               06  
521       248106               07  
522       236053               08  
523       242200               09  
524       213080            

In [127]:
df_song_info_stud_alb['album'].value_counts()

Red (Taylor's Version)              30
Fearless (Taylor's Version)         26
Midnights (The Til Dawn Edition)    23
1989 (Taylor's Version) [Deluxe]    22
Speak Now (Taylor's Version)        22
Red (Deluxe Edition)                22
Speak Now (Deluxe Edition)          20
1989 (Deluxe Edition)               19
Fearless Platinum Edition           19
Lover                               18
evermore (deluxe version)           17
folklore (deluxe version)           17
reputation                          15
Debut                               15
Name: album, dtype: int64

Observations: Successfully renamed Taylor Swift album to be Debut

In [128]:
# Create new column for album id

result = []
for value in df_song_info_stud_alb['album']:
    if value == "Red (Taylor's Version)":
        result.append("REDTV")
    elif value == "Midnights (The Til Dawn Edition)":
        result.append("MIDN")
    elif value == "Red (Deluxe Edition)":
        result.append("RED")
    elif value == "Fearless (Taylor's Version)":
        result.append("FEARTV")
    elif value == "1989 (Taylor's Version) [Deluxe]":
        result.append("1989TV")        
    elif value == "Speak Now (Taylor's Version)":
        result.append("SPEAKTV")
    elif value == "Red (Deluxe Edition)":
        result.append("RED")
    elif value == "Speak Now (Deluxe Edition)":
        result.append("SPEAK")
    elif value == "1989 (Deluxe Edition)":
        result.append("1989")
    elif value == "Fearless Platinum Edition":
        result.append("FEAR")
    elif value == "Lover":
        result.append("LOVE")
    elif value == "evermore (deluxe version)":
        result.append("EVER")
    elif value == "folklore (deluxe version)":
        result.append("FOLK")
    elif value == "reputation":
        result.append("REP")
    elif value == "Debut":
        result.append("DEBUT")
    else:
        result.append("Unknown")

In [129]:
df_song_info_stud_alb['album_id'] = result

C:\Users\leobs\AppData\Local\Temp\ipykernel_14464\3394860859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_song_info_stud_alb['album_id'] = result


In [130]:
df_song_info_stud_alb['album_id'].value_counts(ascending=False)

REDTV      30
FEARTV     26
MIDN       23
1989TV     22
SPEAKTV    22
RED        22
SPEAK      20
1989       19
FEAR       19
LOVE       18
EVER       17
FOLK       17
REP        15
DEBUT      15
Name: album_id, dtype: int64

In [131]:
# Crosstab of albums, by name, and song count to check with album_id counts

crosstab_albums = df_song_info_stud_alb.groupby('album').agg({'song_name':['count']}, dropna=False, ascending =False)

In [132]:
crosstab_albums

song_name
                                     count
album                                     
1989 (Deluxe Edition)                   19
1989 (Taylor's Version) [Deluxe]        22
Debut                                   15
Fearless (Taylor's Version)             26
Fearless Platinum Edition               19
Lover                                   18
Midnights (The Til Dawn Edition)        23
Red (Deluxe Edition)                    22
Red (Taylor's Version)                  30
Speak Now (Deluxe Edition)              20
Speak Now (Taylor's Version)            22
evermore (deluxe version)               17
folklore (deluxe version)               17
reputation                              15

Song count values match album counts for both "album" and actually tracks.

In [133]:
# Rename album to album_title and song to song_title to match other dataframes

In [134]:
df_song_info_stud_alb.rename(columns = {'song_name': 'song_title', 'album': 'album_title'}, inplace=True)

C:\Users\leobs\AppData\Local\Temp\ipykernel_14464\3169183802.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_song_info_stud_alb.rename(columns = {'song_name': 'song_title', 'album': 'album_title'}, inplace=True)


In [135]:
df_song_info_stud_alb.columns

Index(['song_title', 'album_title', 'album_release_date', 'track_number',
       'acousticness', 'danceability', 'energy', 'instrumentalness',
       'liveness', 'loudness', 'speechiness', 'tempo', 'valence',
       'song_popularity', 'duration_ms', 'track_number_str', 'album_id'],
      dtype='object')

In [136]:
df_song_info_stud_alb['track_number_str'].value_counts(dropna=False, ascending=False)

01    14
03    14
04    14
05    14
06    14
07    14
08    14
09    14
10    14
11    14
12    14
13    14
14    14
15    14
02    14
17    12
16    12
18    10
19     9
20     7
21     6
22     6
23     3
24     2
25     2
26     2
27     1
28     1
29     1
30     1
Name: track_number_str, dtype: int64

In [137]:
df_song_info_stud_alb.head()

song_title  \
0         Welcome To New York (Taylor's Version)   
1                 Blank Space (Taylor's Version)   
2                       Style (Taylor's Version)   
3            Out Of The Woods (Taylor's Version)   
4  All You Had To Do Was Stay (Taylor's Version)   

                        album_title album_release_date  track_number  \
0  1989 (Taylor's Version) [Deluxe]         10/27/2023             1   
1  1989 (Taylor's Version) [Deluxe]         10/27/2023             2   
2  1989 (Taylor's Version) [Deluxe]         10/27/2023             3   
3  1989 (Taylor's Version) [Deluxe]         10/27/2023             4   
4  1989 (Taylor's Version) [Deluxe]         10/27/2023             5   

   acousticness  danceability  energy  instrumentalness  liveness  loudness  \
0      0.009420         0.757   0.610          0.000037    0.3670    -4.840   
1      0.088500         0.733   0.733          0.000000    0.1680    -5.376   
2      0.000421         0.511   0.822          0.019700    0.0899    -4.785   
3      0.000537         0.545   0.885          0.000056    0.3850    -5.968   
4      0.000656         0.588   0.721          0.000000    0.1310    -5.579   

   speechiness    tempo  valence  song_popularity  duration_ms  \
0       0.0327  116.998    0.685               80       212600   
1       0.0670   96.057    0.701               80       231833   
2       0.0397   94.868    0.305               81       231000   
3       0.0447   92.021    0.206               80       235800   
4       0.0317   96.997    0.520               79       193289   

  track_number_str album_id  
0               01   1989TV  
1               02   1989TV  
2               03   1989TV  
3               04   1989TV  
4               05   1989TV

In [138]:
# Create song_id column using the track_number and album_id

df_song_info_stud_alb['song_id'] = df_song_info_stud_alb['album_id'] + df_song_info_stud_alb['track_number_str']

C:\Users\leobs\AppData\Local\Temp\ipykernel_14464\1542885659.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_song_info_stud_alb['song_id'] = df_song_info_stud_alb['album_id'] + df_song_info_stud_alb['track_number_str']


In [139]:
df_song_info_stud_alb.head()

song_title  \
0         Welcome To New York (Taylor's Version)   
1                 Blank Space (Taylor's Version)   
2                       Style (Taylor's Version)   
3            Out Of The Woods (Taylor's Version)   
4  All You Had To Do Was Stay (Taylor's Version)   

                        album_title album_release_date  track_number  \
0  1989 (Taylor's Version) [Deluxe]         10/27/2023             1   
1  1989 (Taylor's Version) [Deluxe]         10/27/2023             2   
2  1989 (Taylor's Version) [Deluxe]         10/27/2023             3   
3  1989 (Taylor's Version) [Deluxe]         10/27/2023             4   
4  1989 (Taylor's Version) [Deluxe]         10/27/2023             5   

   acousticness  danceability  energy  instrumentalness  liveness  loudness  \
0      0.009420         0.757   0.610          0.000037    0.3670    -4.840   
1      0.088500         0.733   0.733          0.000000    0.1680    -5.376   
2      0.000421         0.511   0.822          0.019700    0.0899    -4.785   
3      0.000537         0.545   0.885          0.000056    0.3850    -5.968   
4      0.000656         0.588   0.721          0.000000    0.1310    -5.579   

   speechiness    tempo  valence  song_popularity  duration_ms  \
0       0.0327  116.998    0.685               80       212600   
1       0.0670   96.057    0.701               80       231833   
2       0.0397   94.868    0.305               81       231000   
3       0.0447   92.021    0.206               80       235800   
4       0.0317   96.997    0.520               79       193289   

  track_number_str album_id   song_id  
0               01   1989TV  1989TV01  
1               02   1989TV  1989TV02  
2               03   1989TV  1989TV03  
3               04   1989TV  1989TV04  
4               05   1989TV  1989TV05

In [140]:
df_song_info_stud_alb.tail(50)

song_title                 album_title  \
445                                    Dear John  Speak Now (Deluxe Edition)   
446                                         Mean  Speak Now (Deluxe Edition)   
447                              The Story Of Us  Speak Now (Deluxe Edition)   
448                                Never Grow Up  Speak Now (Deluxe Edition)   
449                                    Enchanted  Speak Now (Deluxe Edition)   
450                          Better Than Revenge  Speak Now (Deluxe Edition)   
451                                     Innocent  Speak Now (Deluxe Edition)   
452                                      Haunted  Speak Now (Deluxe Edition)   
453                                    Last Kiss  Speak Now (Deluxe Edition)   
454                                    Long Live  Speak Now (Deluxe Edition)   
455                                         Ours  Speak Now (Deluxe Edition)   
456                          If This Was A Movie  Speak Now (Deluxe Edition)   
457                                     Superman  Speak Now (Deluxe Edition)   
458                  Back To December - Acoustic  Speak Now (Deluxe Edition)   
459                   Haunted - Acoustic Version  Speak Now (Deluxe Edition)   
460                               Mine - POP Mix  Speak Now (Deluxe Edition)   
475                               Jump Then Fall   Fearless Platinum Edition   
476                                  Untouchable   Fearless Platinum Edition   
477             Forever & Always - Piano Version   Fearless Platinum Edition   
478                        Come In With The Rain   Fearless Platinum Edition   
479                                    SuperStar   Fearless Platinum Edition   
480                   The Other Side Of The Door   Fearless Platinum Edition   
481                                     Fearless   Fearless Platinum Edition   
482                                      Fifteen   Fearless Platinum Edition   
483                                   Love Story   Fearless Platinum Edition   
484                                  Hey Stephen   Fearless Platinum Edition   
485                                  White Horse   Fearless Platinum Edition   
486                           You Belong With Me   Fearless Platinum Edition   
487                                      Breathe   Fearless Platinum Edition   
488                                  Tell Me Why   Fearless Platinum Edition   
489                             You're Not Sorry   Fearless Platinum Edition   
490                          The Way I Loved You   Fearless Platinum Edition   
491                             Forever & Always   Fearless Platinum Edition   
492                                 The Best Day   Fearless Platinum Edition   
493                                       Change   Fearless Platinum Edition   
515                                   Tim McGraw                       Debut   
516                              Picture To Burn                       Debut   
517  Teardrops On My Guitar - Radio Single Remix                       Debut   
518                        A Place in this World                       Debut   
519                                  Cold As You                       Debut   
520                                  The Outside                       Debut   
521                   Tied Together with a Smile                       Debut   
522                               Stay Beautiful                       Debut   
523                            Should've Said No                       Debut   
524                    Mary's Song (Oh My My My)                       Debut   
525                                     Our Song                       Debut   
526                I'm Only Me When I'm With You                       Debut   
527                                    Invisible                       Debut   
528                       A Perfectly Good Heart                       Debut   
529         Teardrops on My Guitar - P

Succesfully added new column with song id, which can be matched to other dataframes for analysis.

## Final song_info stats and value counts

In [141]:
df_song_info_stud_alb.describe()

track_number  acousticness  danceability      energy  instrumentalness  \
count    285.000000    285.000000    285.000000  285.000000        285.000000   
mean      11.080702      0.279533      0.588951    0.587849          0.002758   
std        6.592020      0.310108      0.109833    0.183155          0.024231   
min        1.000000      0.000191      0.292000    0.118000          0.000000   
25%        6.000000      0.026800      0.516000    0.469000          0.000000   
50%       11.000000      0.120000      0.595000    0.608000          0.000001   
75%       16.000000      0.525000      0.654000    0.732000          0.000027   
max       30.000000      0.971000      0.897000    0.950000          0.328000   

         liveness    loudness  speechiness       tempo     valence  \
count  285.000000  285.000000   285.000000  285.000000  285.000000   
mean     0.147512   -7.205772     0.060321  124.349042    0.405238   
std      0.091840    2.868390     0.086982   30.951337    0.198153   
min      0.035700  -17.932000     0.023100   68.534000    0.038200   
25%      0.094800   -8.796000     0.030200   99.953000    0.248000   
50%      0.116000   -6.639000     0.036500  121.009000    0.395000   
75%      0.167000   -5.191000     0.055800  145.880000    0.539000   
max      0.657000   -1.909000     0.912000  208.918000    0.942000   

       song_popularity    duration_ms  
count       285.000000     285.000000  
mean         68.796491  240297.866667  
std          12.541508   47181.045343  
min          39.000000  107133.000000  
25%          62.000000  212600.000000  
50%          71.000000  235466.000000  
75%          78.000000  261230.000000  
max         100.000000  613026.000000

In [142]:
df_song_info_stud_alb.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 285 entries, 0 to 529
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   song_title          285 non-null    object 
 1   album_title         285 non-null    object 
 2   album_release_date  285 non-null    object 
 3   track_number        285 non-null    int64  
 4   acousticness        285 non-null    float64
 5   danceability        285 non-null    float64
 6   energy              285 non-null    float64
 7   instrumentalness    285 non-null    float64
 8   liveness            285 non-null    float64
 9   loudness            285 non-null    float64
 10  speechiness         285 non-null    float64
 11  tempo               285 non-null    float64
 12  valence             285 non-null    float64
 13  song_popularity     285 non-null    int64  
 14  duration_ms         285 non-null    int64  
 15  track_number_str    285 non-null    object 
 16  album_id

In [145]:
df_song_info_stud_alb.shape

(285, 18)

In [146]:
df_song_info_stud_alb['album_title'].value_counts(dropna=False)

Red (Taylor's Version)              30
Fearless (Taylor's Version)         26
Midnights (The Til Dawn Edition)    23
1989 (Taylor's Version) [Deluxe]    22
Speak Now (Taylor's Version)        22
Red (Deluxe Edition)                22
Speak Now (Deluxe Edition)          20
1989 (Deluxe Edition)               19
Fearless Platinum Edition           19
Lover                               18
evermore (deluxe version)           17
folklore (deluxe version)           17
reputation                          15
Debut                               15
Name: album_title, dtype: int64

In [147]:
df_song_info_stud_alb['album_release_date'].value_counts(dropna=False)

11/12/2021    30
4/9/2021      26
5/26/2023     23
10/27/2023    22
7/7/2023      22
10/22/2012    22
10/25/2010    20
10/27/2014    19
11/11/2008    19
8/23/2019     18
1/7/2021      17
8/18/2020     17
11/10/2017    15
10/24/2006    15
Name: album_release_date, dtype: int64

In [148]:
df_song_info_stud_alb['track_number_str'].value_counts(dropna=False)

01    14
03    14
04    14
05    14
06    14
07    14
08    14
09    14
10    14
11    14
12    14
13    14
14    14
15    14
02    14
17    12
16    12
18    10
19     9
20     7
21     6
22     6
23     3
24     2
25     2
26     2
27     1
28     1
29     1
30     1
Name: track_number_str, dtype: int64

In [149]:
df_song_info_stud_alb['album_id'].value_counts(dropna=False)

REDTV      30
FEARTV     26
MIDN       23
1989TV     22
SPEAKTV    22
RED        22
SPEAK      20
1989       19
FEAR       19
LOVE       18
EVER       17
FOLK       17
REP        15
DEBUT      15
Name: album_id, dtype: int64

In [150]:
song_id_unique_count = df_song_info_stud_alb['song_id'].nunique()

In [151]:
song_id_unique_count

285

Cleaned dataframe exported as song_info_clean

# 04. Dataframe exports

In [152]:
#Export subset dataframe

df_song_info_no_use_albums.to_csv(os.path.join(path, '02 Manipulated Data', 'song_info_no_use_albums.csv'))

In [153]:
#Export gsearch_st cleaned df

df_gsearch_st.to_csv(os.path.join(path, '02 Manipulated Data', 'gsearch_st_clean.csv'))

In [154]:
#Export song_info cleaned df

df_song_info_stud_alb.to_csv(os.path.join(path, '02 Manipulated Data', 'song_info_clean.csv'))